compiler/environment.ml

```diff
@@ -0,0 +1,116 @@
+open Ast;;
+
+module FunctionMap = Map.Make(String);;
+module VariableMap = Map.Make(String);;
+module ArrayTypeMap = Map.Make(String);;
+
+exception VarAlreadyDeclared;;
+exception VarNotDeclared;;
+exception FunctionAlreadyDeclared;;
+exception FunctionNotDeclared;;
+exception IncorrectFunctionParameterTypes;;
+exception MixedTypeArray;;
+exception ArrayLiteralTypeMismatch;;
+
+type func_info  = {
+  id : string; 
+  return : data_type; 
+  args : data_type list;
+  arg_names: string list;
+}
+
+type symbol_table = {
+  func_map: func_info FunctionMap.t;
+  var_map: data_type VariableMap.t;
+  array_type_map: data_type ArrayTypeMap.t;
+}
+
+let create = 
+  {
+    func_map = FunctionMap.empty;
+    var_map = VariableMap.empty;
+    array_type_map = ArrayTypeMap.empty;
+  }
+
+let update f_map v_map a_type_map =
+  {
+    func_map = f_map;
+    var_map = v_map;
+    array_type_map = a_type_map;
+  }
+
+let string_to_data_type (s : string) = match s
+  with "int" -> Int
+  | "float" -> Float
+  | "bool" -> Bool
+  | "string" -> String
+  | "array" -> Array(Int)
+  | "json" -> Json
+  | _ -> raise (Failure "unsupported data type")
+
+let declare_var (id : string) (data_type : string) (env : symbol_table) =
+  if VariableMap.mem id env.var_map then 
+    raise VarAlreadyDeclared
+  else 
+    let update_var_map = VariableMap.add id (string_to_data_type(data_type)) env.var_map in
+    update env.func_map update_var_map env.array_type_map
+
+let var_type (id : string) (env : symbol_table) =
+  if VariableMap.mem id env.var_map then
+    VariableMap.find id env.var_map
+  else
+    raise VarNotDeclared
+
+let create_func (func_name: string) (ret_type : string) (args : arg_decl list) =
+  {
+    id = func_name;
+    return = (string_to_data_type ret_type);
+    args = List.map (fun arg -> string_to_data_type arg.var_type) args;
+    arg_names = List.map (fun arg -> arg.var_name) args;
+  }
+
+let define_array_type (args : data_type list) (env : symbol_table) (id : string) =
+  if List.length args != 0 then
```

Comment:
What is the equivalent Java code for "array a = []" ? It gets approved by our parser and is handled in line 79, but by Java standards, an array initialization must always be accompanied by the array type. Imagine this code:
```
array a = []
if (randomcondition) {
a[0] = 1
}
else {
a[0] = "yo"
}
```
There is no way to even know what the type of the array will be at compile time because we don't know the value of randomcondition, and we can't generate Java code because we have to provide that type.

The solution is to disallow instantiating an array without an array type. This requires some change to the parser so that we can do something like array `a = int`, or `array a = [1]`, but not `array a = []`.